go/ct/common/hashes256.go

```diff
@@ -0,0 +1,114 @@
+//
+// Copyright (c) 2024 Fantom Foundation
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file and at fantom.foundation/bsl11.
+//
+// Change Date: 2028-4-16
+//
+// On the date above, in accordance with the Business Source License, use
+// of this software will be governed by the GNU Lesser General Public Licence v3
+//
+
+package common
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/Fantom-foundation/Tosca/go/vm"
+	"pgregory.net/rand"
+)
+
+// ImmutableHashArray is an immutable array of 256 vm.Hash that can be trivially cloned/compared.
+type ImmutableHashArray struct {
+	data *[256]vm.Hash
+}
+
+func NewImmutableHashArray(hashes ...vm.Hash) ImmutableHashArray {
+	var data [256]vm.Hash
+	copy(data[:], hashes)
+	return ImmutableHashArray{data: &data}
+}
+
+func (b ImmutableHashArray) Equal(other ImmutableHashArray) bool {
+	if b.data == other.data {
+		return true
+	}
+	if b.data == nil || other.data == nil {
+		return false
+	}
+	for i := 0; i < 256; i++ {
+		if b.data[i] != other.data[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func RandomImmutableHashArray(rnd *rand.Rand) ImmutableHashArray {
+	hashes := ImmutableHashArray{}
+	hashes.data = new([256]vm.Hash)
+	for i := 0; i < 256; i++ {
+		hashes.data[i] = GetRandomHash(rnd)
+	}
+	return hashes
+}
+
+func (b ImmutableHashArray) ToBytes() []byte {
+	bytes := []byte{}
+	for i := 0; i < 256; i++ {
+		bytes = append(bytes, b.data[i][:]...)
+	}
+	return bytes
+}
+
+func (b ImmutableHashArray) FromBytes(bytes []byte) ImmutableHashArray {
+	hashes := ImmutableHashArray{}
+	for i := 0; i < 256; i++ {
+		copy(hashes.data[i][:], bytes[i*32:(i+1)*32])
+	}
+	return hashes
+}
+
+func (b ImmutableHashArray) ToHash() [256]vm.Hash {
+	return [256]vm.Hash(*b.data)
+}
+
+func (b ImmutableHashArray) String() string {
+	return fmt.Sprintf("%x", b.data)
+}
+
+func (b ImmutableHashArray) MarshalJSON() ([]byte, error) {
+	return json.Marshal(b.data)
+}
+
+// Implement the UnmarshalJSON method for ImmutableHashArray
```

Comment:
The comment is not formatted according to the general Go convention and does not provide any information that would not be obvious from the existence of this function.